google-datacatalog-kafka-connector/src/google/datacatalog_connectors/kafka/scrape/metadata_scraper.py

```diff
@@ -1,37 +1,117 @@
 import logging
+
+import confluent_kafka
+from confluent_kafka.admin import ConfigResource
+from confluent_kafka.cimpl import KafkaException
+
 from google.datacatalog_connectors.kafka.config.\
     metadata_constants import MetadataConstants
+from .metadata_values_converter import MetadataValuesConverter
 
 
 class MetadataScraper:
 
-    def __init__(self, consumer):
-        self._consumer = consumer
+    def __init__(self, client, bootstrap_server):
+        self._adminClient = client
+        self._bootstrap_server = bootstrap_server
 
     def get_metadata(self):
         try:
-            raw_metadata = self._consumer.list_topics(timeout=20)
-            topic_metadata = self._get_topic_metadata(raw_metadata)
+            raw_metadata = self._adminClient.list_topics(timeout=20)
+            topics_metadata = self._get_topics_metadata(raw_metadata)
             cluster_metadata = self._get_cluster_metadata(raw_metadata)
-            cluster_metadata.update(topic_metadata)
+            cluster_metadata.update(topics_metadata)
             return cluster_metadata
         except:  # noqa:E722 silence linter complaint about bare except
             logging.error(
                 'Error connecting to the system to extract metadata.')
             raise
-        finally:
-            self._consumer.close()
-
-    def _get_topic_metadata(self, metadata_object):
-        topic_names = metadata_object.topics.keys()
-        topic_metadata = {MetadataConstants.TOPICS: topic_names}
-        return topic_metadata
 
     def _get_cluster_metadata(self, metadata_object):
         cluster_id = metadata_object.cluster_id
         num_brokers = len(metadata_object.brokers)
         cluster_metadata = {
             MetadataConstants.CLUSTER_ID: cluster_id,
-            MetadataConstants.BROKERS_NUM: num_brokers
+            MetadataConstants.NUM_BROKERS: num_brokers,
+            MetadataConstants.BOOTSTRAP_SERVER: self._bootstrap_server
         }
         return cluster_metadata
+
+    def _get_topics_metadata(self, metadata_object):
+        topic_names = metadata_object.topics.keys()
+        descriptions = list()
+        config_resources = [
+            ConfigResource(confluent_kafka.admin.RESOURCE_TOPIC, topic_name)
+            for topic_name in topic_names
+        ]
+        config_futures = self._adminClient.describe_configs(config_resources,
+                                                            request_timeout=10)
+        for topic, future in config_futures.items():
+            try:
+                config = future.result()
```

Comment:
According to the [doc](https://docs.confluent.io/current/clients/confluent-kafka-python/#confluent_kafka.admin.AdminClient), it does return one future for each topic. Could you elaborate on what you have in mind when talking about parallelizing the future fetching?

Reply:
The `future.result()` in the iteration loop is working as serial operations, being called for each topic, so each time it waits for the return to go to the next.

I'd add some logging to see how much time is spent there, so see if it is worth optimizing that.

You could combine batches of futures and parallelize that to fetch the results. But would only spent time evaluating this, if the amount of time is considerable. 

Since a Kafka cluster won't have that many topics to make it worth.

To sum up, for now I would suggest just adding the logging with the elapsed time spent on this.

Reply:
Okay, I see! sure, let's log it.

Reply:
I have measured it now, and description of 1000 topics takes approx 1.2 sec for me. This information is now also being logged in logging.info()